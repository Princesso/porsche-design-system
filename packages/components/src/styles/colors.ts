import type { Theme } from '@porsche-design-system/utilities-v2';
import { isThemeDark } from '../utils';

type ThemedColor =
  | 'primaryColor'
  | 'primaryColorDarken'
  | 'backgroundColor'
  | 'backgroundSurfaceColor'
  | 'contrastLowColor'
  | 'contrastMediumColor'
  | 'contrastHighColor'
  | 'contrastHighColorDarken'
  | 'hoverColor'
  | 'hoverColorDarken'
  | 'activeColor'
  | 'focusColor'
  | 'disabledColor'
  | 'errorColor'
  | 'errorColorDarken'
  | 'errorSoftColor'
  | 'successColor'
  | 'successColorDarken'
  | 'successSoftColor'
  | 'warningColor'
  | 'warningSoftColor'
  | 'infoColor'
  | 'infoSoftColor';

export type ThemedColors = { [key in ThemedColor]: string };

/* Auto Generated Start */
const themeLight = {
  primaryColor: '#010205',
  primaryColorDarken: '#000000',
  backgroundColor: '#FFF',
  backgroundSurfaceColor: '#EEEFF2',
  contrastLowColor: '#D8D8DB',
  contrastMediumColor: '#949598',
  contrastHighColor: '#535457',
  contrastHighColorDarken: '#353638',
  hoverColor: '#D8D8DB',
  hoverColorDarken: '#b8b8be',
  activeColor: '#D8D8DB',
  focusColor: '#0A0AFF',
  disabledColor: '#D8D8DB',
  errorColor: '#E7323B',
  errorColorDarken: '#c51720',
  errorSoftColor: '#FCE8E9',
  successColor: '#32B85B',
  successColorDarken: '#258843',
  successSoftColor: '#EBFAF0',
  warningColor: '#FECC1B',
  warningSoftColor: '#FFF9E6',
  infoColor: '#1E5BEB',
  infoSoftColor: '#E8EEFD'
};

const themeDark = {
  primaryColor: '#FBFCFF',
  primaryColorDarken: '#beceff',
  backgroundColor: '#0E0E12',
  backgroundSurfaceColor: '#212225',
  contrastLowColor: '#404044',
  contrastMediumColor: '#7E7F82',
  contrastHighColor: '#AFB0B3',
  contrastHighColorDarken: '#909195',
  hoverColor: '#404044',
  hoverColorDarken: '#222224',
  activeColor: '#404044',
  focusColor: '#0A0AFF',
  disabledColor: '#404044',
  errorColor: '#CB3333',
  errorColorDarken: '#9a2727',
  errorSoftColor: '#290A0A',
  successColor: '#00C77E',
  successColorDarken: '#008a57',
  successSoftColor: '#003320',
  warningColor: '#DDB84B',
  warningSoftColor: '#2B2208',
  infoColor: '#027FFC',
  infoSoftColor: '#001A33'
};

const themes = {
  'light': themeLight,
  'dark': themeDark
};
/* Auto Generated End */

export const getThemedColors = (theme: Theme): ThemedColors => {
  return themes[theme];
};

export const getInvertedThemedColors = (theme: Theme): ThemedColors => {
  return getThemedColors(isThemeDark(theme) ? 'light' : 'dark');
};
