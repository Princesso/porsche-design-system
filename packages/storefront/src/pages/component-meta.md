<template>
  <p-table>
    <p-table-head>
      <p-table-head-row v-html="headRow"></p-table-head-row>
    </p-table-head>
    <p-table-body v-html="body"></p-table-body>
  </p-table>
</template>

<script lang="ts">
import Vue from 'vue';  
import Component from 'vue-class-component';
import { getComponentMeta, TAG_NAMES } from '@porsche-design-system/shared';
import type { ComponentMeta } from '@porsche-design-system/shared';

@Component
export default class Code extends Vue {
  get headRow(): string {
    return ['', ...TAG_NAMES].map((tagName) => {
      const { isDeprecated } = getComponentMeta(tagName) || {};
      return `<p-table-head-cell>${tagName}${isDeprecated ? ' ðŸš«' : ''}</p-table-head-cell>`;
    }).join('');
  }

  get body(): string {
    const rowKeys: (keyof ComponentMeta)[] = [
      'isDelegatingFocus',
      'isThemeable',
      'props',
      'requiredProps',
      'eventNames',
      'namedSlots',
      'nestedComponents',
      'hasSlottedCss',
    ];

    const content = rowKeys.map(key => {
      const cells = TAG_NAMES.map(tagName => {
        const meta = getComponentMeta(tagName);
        let value = meta[key];

        if (value && (key === 'props' || key === 'eventNames')) {
          const { deprecatedProps = [], deprecatedEventNames = [] } = meta;

          if (key === 'props') {
            value = Object.keys(value);
            value = value.map(val => deprecatedProps.includes(val) ? `${val} ðŸš«` : val);
          } else if (key === 'eventNames') {
            value = value.map(val => deprecatedEventNames.includes(val) ? `${val} ðŸš«` : val);
          }
        }

        let cellContent = value 
          ? Array.isArray(value)
            ? value.sort().map(val => key === 'nestedComponents' ? val : `<code>${val}</code>`).join('<br>') 
            : value
          : '';
        cellContent = cellContent === true ? 'âœ…' : cellContent;

        return `<p-table-cell>${cellContent}</p-table-cell>`;
      }).join('');

      return `<p-table-row>
  <p-table-cell>${key}</p-table-cell>
  ${cells}
</p-table-row>`;
    }).join('');
    
    return content;
  }
}
</script>

<style scoped lang="scss">
  @import '~@porsche-design-system/components-js/styles/scss';

  p-table {
    max-width: none !important;
  }

  :deep(p-table-cell) {
    vertical-align: top;
  }

  :deep(code) {
    @include pds-text-x-small;
    font-family: Menlo, Monaco, Consolas, 'Courier New', monospace;
    padding: 0.125rem $pds-spacing-static-small;
    background-color: mix($pds-theme-light-primary, $pds-theme-light-background-base, 10%);
    border-radius: 3px;
    color: $pds-theme-light-primary;
  }
</style>
