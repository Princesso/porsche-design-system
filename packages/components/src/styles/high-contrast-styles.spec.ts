import type { TagName } from '@porsche-design-system/shared';
import { TAG_NAMES } from '@porsche-design-system/shared';
import { getComponentMeta } from '@porsche-design-system/component-meta';
import {
  addParentAndSetRequiredProps,
  componentFactory,
  getComponentCssObject,
  getComponentCssSpy,
} from '../test-utils';
import * as a11yUtils from '../utils/a11y/a11y';
import { getHighContrastColors } from './colors';

const tagNamesWithJss = TAG_NAMES.filter((tagName) => {
  const meta = getComponentMeta(tagName);
  return !meta.isInternal && meta.styling === 'jss';
});

it.each<TagName>(tagNamesWithJss)('should have only high contrast styles for %s', (tagName) => {
  // mock to get the result from getComponentCss() directly
  const spy = getComponentCssSpy();

  const component = componentFactory(tagName);

  Object.defineProperty(a11yUtils, 'isHighContrastMode', { value: true });

  // some components like grid-item and text-list-item require a parent to apply styles
  // some components require a parent and certain props in order to work
  addParentAndSetRequiredProps(tagName, component);

  component.render();
  expect(spy).toBeCalledTimes(1);

  const highContrastColors = getHighContrastColors();
  const cssObject = getComponentCssObject(spy);
  const filteredCSS = [];

  const findKey = (obj: Record<string, any>, val: string): Record<string, any> | null => {
    if (typeof obj === 'object') {
      for (const [parentKey, value] of Object.entries(obj)) {
        if (typeof value === 'object') {
          const match = Object.entries(value).find(([, childValue]) => childValue === val);
          if (match) {
            const [childKey, childValue] = match;
            return { [parentKey]: { [childKey]: childValue } };
          }
        }
      }
    }
    return null;
  };

  Object.values(highContrastColors).forEach((value) => {
    const res = findKey(cssObject, value);
    if (res !== null) {
      filteredCSS.push(res);
    }
    return null;
  });
  filteredCSS.length && expect(filteredCSS).toMatchSnapshot();
});
