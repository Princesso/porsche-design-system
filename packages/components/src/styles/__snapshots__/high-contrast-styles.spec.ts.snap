// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have only high contrast styles for p-button 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-carousel 1`] = `
[
  {
    "background": "CanvasText",
  },
]
`;

exports[`should have only high contrast styles for p-divider 1`] = `
[
  {
    "background": "CanvasText",
  },
]
`;

exports[`should have only high contrast styles for p-link 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-link-social 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-popover 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-scroller 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-segmented-control-item 1`] = `
[
  {
    "border-color": "Highlight",
  },
]
`;

exports[`should have only high contrast styles for p-spinner 1`] = `
[
  {
    "stroke": "Canvas",
  },
  {
    "stroke": "CanvasText",
  },
]
`;

exports[`should have only high contrast styles for p-switch 1`] = `
[
  {
    "background-color": "CanvasText",
  },
]
`;

exports[`should have only high contrast styles for p-tabs-bar 1`] = `
[
  {
    "background": "CanvasText",
  },
]
`;

exports[`should have only high contrast styles for p-wordmark 1`] = `
[
  {
    "fill": "CanvasText",
  },
]
`;
